Documentation/Common-Gotchas.md

```diff
@@ -0,0 +1,232 @@
+are within the assumed limitations of `curry`. But, it is also less clear how we
+should write `User.decode`. How can we extract the 3 avatar fields from `JSON`
+to build `User.Avatar` first, and then use that to build `User`? The key insight
+is to see that `j <| "key"` returns a `Decoded<T>` value, representing one step
+of decoding. So, we can decode a `User.Avatar` first using `j`, and then just
+plug that directly into `curry(User.init)`:
+
+```swift
+extension User.Avatar: Decodable {
+  static func decode(j: JSON) -> Decoded<User.Avatar> {
+    return curry(User.Avatar.init)
+      <^> j <| "small_avatar"
+      <*> j <| "medium_avatar"
+      <*> j <| "large_avatar"
+  }
+}
+
+extension User: Decodable {
+  static func decode(j: JSON) -> Decoded<User> {
+    return curry(User.init)
+      <^> j <| "id"
+      <*> j <| "name"
+      <*> j <| "bio"
+      <*> User.Avatar.decode(j)
+  }
+}
+```
+
+We have now successfully gotten around `curry`’s limitations, but also produced
+a better, normalized model! We also retain all of the great error handling that
+Argo provides, so if there is a decoding error in `User.Avatar` it will
+propogate through to a decoding error in `User`.
+
+## Compilation errors
+
+Swift’s compile errors can be mystifying if you are not intimately familiar with
+the internals of Argo. Here we describe some of the most common errors and how
+to fix them.
+
+### Complex expressions
+
+Considering that the implementation of `decode` on a model consists of fully
+currying a method and composing it with many small pieces, each of which may be
+the composition of pieces, it is not surprising that Swift sometimes has
+difficulty compiling the expression in a reasonable amount of time. This
+manifests itself with the following error:
+
+> Expression was too complex to be solved in reasonable time; consider breaking
+up the expression into distinct sub-expressions.
+
+Fortunately this is easy to solve. A first step is to simply store
+`curry(Model.init)` in a variable instead of using it directly:
+
+```swift
+static func decode(j: JSON) -> Decoded<Model> {
+  let create = Curry(Model.init)
+  return create
+    <^> j <| "key1"
+    <*> j <| "key2"
+    ...
+```
+
+Many times this can fix the problem. If you still have complex expression
+issues, the next step is to pick a midpoint in your decoding chain and set an
+intermediate variable:
+
+```swift
+static func decode(j: JSON) -> Decoded<Model> {
+  let create = Curry(Model.init)
+  let tmp = create
+    <^> j <| "key_a_1"
+    <*> j <| "key_a_2"
+    ...
+  return tmp
+    <*> j <| "key_b_1"
+    <*> j <| "key_b_2"
+    ...
+```
+
+This should almost surely fix all of your complex expression issues, but if not,
+be prepared to try one more intermediate variable. Take careful note that in
+defining `tmp` we used the map operator `create <^>` but in the return statement
+we used the applicative operator `tmp <*>`. The map operator is _only_ used with
+the fully curried `init`.
```

Comment:
It appears that there is another solution here. One could also give the type inference of the compiler a break and type all the expressions. Here is an example from the `TestModel` class in our tests.

```swift
extension TestModel: Decodable {
  static func decode(_ json: JSON) -> Decoded<TestModel> {
    return curry(self.init)
      <^> json <| "numerics"                                    as Decoded<TestModelNumerics>
      <*> json <| ["user_opt", "name"]                          as Decoded<String>
      <*> json <| "bool"                                        as Decoded<Bool>
      <*> json <|| "string_array"                               as Decoded<[String]>
      <*> json <||? "string_array_opt"                          as Decoded<[String]?>
      <*> json <|| ["embedded", "string_array"]                 as Decoded<[String]>
      <*> json <||? ["embedded", "string_array_opt"]            as Decoded<[String]?>
      <*> json <|? "user_opt"                                   as Decoded<User?>
      <*> (json <| "dict" >>- { [String: String].decode($0) })  as Decoded<[String: String]>
      <*> json <| "numerics"                                    as Decoded<TestModelNumerics>
      <*> json <| ["user_opt", "name"]                          as Decoded<String>
      <*> json <| "bool"                                        as Decoded<Bool>
      <*> json <|| "string_array"                               as Decoded<[String]>
      <*> json <||? "string_array_opt"                          as Decoded<[String]?>
      <*> json <|| ["embedded", "string_array"]                 as Decoded<[String]>
      <*> json <||? ["embedded", "string_array_opt"]            as Decoded<[String]?>
      <*> json <|? "user_opt"                                   as Decoded<User?>
      <*> (json <| "dict" >>- { [String: String].decode($0) })  as Decoded<[String: String]>
      <*> json <| "numerics"                                    as Decoded<TestModelNumerics>
      <*> json <| ["user_opt", "name"]                          as Decoded<String>
      <*> json <| "bool"                                        as Decoded<Bool>
  }
}
```

This way is arguably uglier but still works without intermediate variables. It can work up to and probably past the limit of curry.